[ignition-common3] Add new port 🤖

traversaro opened the pull request:
- What does your PR fix? 
  - This PR adds a new port for the new major  version of the [ignition-common](https://ignitionrobotics.org/libs/common) library. 

- Which triplets are supported/not supported?
  - All tested triplet (for which the dependency are actually available) should be supported.

- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  - I hope.

As discussed in https://github.com/microsoft/vcpkg/pull/7781, different major version of ignition robotics libraries (https://ignitionrobotics.org/) can be installed side by side, so each new major version is added as a new port. 

In particular, ignition-common3 is a dependency for Gazebo 11 (http://gazebosim.org/blog/gazebo11) and for Ignition Robotics Citadel (that contains Ignition Gazebo 3, see https://www.openrobotics.org/blog/2019/12/11/ignition-citadel-released).

traversaro committed:
[ignition-common3] Add new port

PhoebeHui reviewed (approved):
LGTM, thanks for the PR!

PhoebeHui commented on ports/ignition-common3/CONTROL:
```diff
@@ -0,0 +1,4 @@
+Source: ignition-common3
+Version: 3.5.0
+Build-Depends: dlfcn-win32 (windows|uwp), ffmpeg, freeimage, gts, ignition-cmake2, ignition-math6, ignition-modularscripts, tinyxml2
```
Could you also add Homepage?

traversaro replied:
Done in https://github.com/microsoft/vcpkg/pull/11273/commits/96416df8eababf5732d7af50e8a540e2be985496 .

PhoebeHui commented:
@traversaro, thanks for quick updates!

I noticed it failed on linux CI testing, could you have a look?


CMake Error at /mnt/1/s/installed/x64-linux/share/cmake/ignition-cmake2/cmake2/IgnConfigureBuild.cmake:66 (message):
  Errors encountered in build.  Please see BUILD ERRORS above.
Call Stack (most recent call first):
  CMakeLists.txt:120 (ign_configure_build)


[config-x64-linux-dbg-err.log](https://github.com/microsoft/vcpkg/files/4614637/config-x64-linux-dbg-err.log)
[config-x64-linux-dbg-out.log](https://github.com/microsoft/vcpkg/files/4614638/config-x64-linux-dbg-out.log)

Electric-Turtle commented:
Something I have noticed:

This port doesn't properly install the cmake config files for ignition-common3's various _subcomponents_, ( "graphics", "profiler", etc). I.e.  the various <subcomponentname>-config.cmake and <subcomponentname>-config-version.cmake never make it to the vcpkg installation directory. However, the actual library files do get built and installed.

for example:
 C:\path\to\vcpkginstalldir\lib\ignition-common3-graphics.lib is present after the install finishes, but the corresponding cmake config file that tells cmake where to find it does not. 
I.e.  C:\path\to\vcpkginstalldir\share\ignition-common3\ignition-common3-graphics-config.cmake _doesn't_ exist.  If you try to do a find_package for ignition-common3 with graphics specified as a component such as the following:
find_package(ignition-common3 CONFIG REQUIRED
   COMPONENTS
     graphics
    )
The find_package call will fail, even though the library was technically built.  For example, if you tried to build Gazebo11 (which I strongly sense is the motivation for this port), the config stage will not work.  The cmake configuration of Gazebo will THINK ignition-common3-graphics doesn't exist, because the config file for it is nowhere to be found.

traversaro commented:
> The find_package call will fail, even though the library was technically built. For example, if you tried to build Gazebo11 (which I strongly sense is the motivation for this port), the config stage will not work. The cmake configuration of Gazebo will THINK ignition-common3-graphics doesn't exist, because the config file for it is nowhere to be found.

Thanks for the comment. For the  moment I am actually building Gazebo11 by building all the ignition related dependencies with colcon (see https://github.com/robotology/robotology-superbuild-dependencies-vcpkg/blob/v0.3.0/.github/workflows/ci.yml#L171), but indeed the  main motivation of brinding all the libraries to vcpkg is to be able to avoid to rely on colcon.

Electric-Turtle commented:
> Thanks for the comment. For the moment I am actually building Gazebo11 by building all the ignition related dependencies with colcon (see https://github.com/robotology/robotology-superbuild-dependencies-vcpkg/blob/v0.3.0/.github/workflows/ci.yml#L171), but indeed the main motivation of brinding all the libraries to vcpkg is to be able to avoid to rely on colcon.

Haha I am currently doing the exact same thing and would also like to avoid having to maintain a custom build of Gazebo with colcon.

traversaro commented:
> 
> 
> @traversaro, thanks for quick updates!
> 
> I noticed ignition-common3:x64-linux failed on linux CI testing, could you have a look?
> 
> CMake Error at /mnt/1/s/installed/x64-linux/share/cmake/ignition-cmake2/cmake2/IgnConfigureBuild.cmake:66 (message):
> Errors encountered in build. Please see BUILD ERRORS above.
> Call Stack (most recent call first):
> CMakeLists.txt:120 (ign_configure_build)
> 
> [config-x64-linux-dbg-err.log](https://github.com/microsoft/vcpkg/files/4614637/config-x64-linux-dbg-err.log)
> [config-x64-linux-dbg-out.log](https://github.com/microsoft/vcpkg/files/4614638/config-x64-linux-dbg-out.log)

I pushed an updated version of the branch that should fix it by backporting the ignition-cmake PR  https://github.com/ignitionrobotics/ign-cmake/pull/95 .

traversaro commented:
> Something I have noticed:

This has been fixed in https://github.com/microsoft/vcpkg/pull/11275 , and the fix will automatically work also for the `ignition-common3` port as soon as that PR is merged.

traversaro commented:
Hi @PhoebeHui , as I see this PR listed under "requires:author-response", if there is anything else that needs my input feel free to ask, thanks!

seanyen commented:
wow, looks like we are about to get all deps for Gazebo11. 😉

traversaro commented:
PR rebased over latest master, and all the changes requested by the reviewer have been addressed.

traversaro commented:
There are  a few failures in the `x64-linux` triplet, in particular it is not able to find anymore uuid, FreeImage and gts : 
~~~
-- BUILD WARNINGS
-- 	Cannot build component [graphics] - Missing: FreeImage
-- 	Cannot build component [graphics] - Missing: gts - GNU Triangulation Surface library
-- END BUILD WARNINGS

-- BUILD ERRORS: These must be resolved before compiling.
-- 	Missing: uuid
-- END BUILD ERRORS

CMake Error at /mnt/vcpkg-ci/install/x64-linux/share/cmake/ignition-cmake2/cmake2/IgnConfigureBuild.cmake:66 (message):
  Errors encountered in build.  Please see BUILD ERRORS above.
Call Stack (most recent call first):
  CMakeLists.txt:120 (ign_configure_build)
~~~

I guess this is probably due to the fact that now `pkg-config` is installed in CI machines, and the `.pc` files installed by those ports are not correct/well formed.

traversaro commented:
> 
> 
> There are a few failures in the `x64-linux` triplet, in particular it is not able to find anymore uuid, FreeImage and gts :
> 
> ```
> -- BUILD WARNINGS
> -- 	Cannot build component [graphics] - Missing: FreeImage
> -- 	Cannot build component [graphics] - Missing: gts - GNU Triangulation Surface library
> -- END BUILD WARNINGS
> 
> -- BUILD ERRORS: These must be resolved before compiling.
> -- 	Missing: uuid
> -- END BUILD ERRORS
> 
> CMake Error at /mnt/vcpkg-ci/install/x64-linux/share/cmake/ignition-cmake2/cmake2/IgnConfigureBuild.cmake:66 (message):
>   Errors encountered in build.  Please see BUILD ERRORS above.
> Call Stack (most recent call first):
>   CMakeLists.txt:120 (ign_configure_build)
> ```
> 
> I guess this is probably due to the fact that now `pkg-config` is installed in CI machines, and the `.pc` files installed by those ports are not correct/well formed.

Actually there was a missing dependency in the `CONTROL` file. This should have been fixed in the last commit.

traversaro commented:
I also backported https://github.com/ignitionrobotics/ign-common/pull/71 to avoid having Windows failures.

traversaro commented:
The linux failure is due to https://github.com/microsoft/vcpkg/issues/11817, so for now I just updated the baseline.

PhoebeHui commented:
/azp run

traversaro commented:
> 
> 
> Hi @PhoebeHui , as I see this PR listed under "requires:author-response", if there is anything else that needs my input feel free to ask, thanks!

Hi @PhoebeHui , friendly ping. : ) 
On which topic do you need a response? Thanks in advance!

traversaro commented:
I rebased the PR over latest master to test the modification with the latest vcpkg.

traversaro commented:
As the `x64-linux` build were failing with: 
~~~
CMake Error in av/src/CMakeLists.txt:
  Imported target "SWSCALE::SWSCALE" includes non-existent path

    "/mnt/vcpkg-ci/packages/ffmpeg_x64-linux/debug/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.
~~~
that was caused by the fact that `ffmpeg` install  non-relocatable .pc files, I also added a commit with a fix for that (related issue: https://github.com/microsoft/vcpkg/issues/11087).

traversaro commented:
The invocation of vcpkg_fixup_pkgconfig on ffmpeg fails with: 
~~~
CMake Error at scripts/cmake/vcpkg_fixup_pkgconfig.cmake:150 (message):
  Found ole32.lib user32.lib and no rule to analyse the flag! Please check
  the *.pc file
~~~
@Neumann-A do you think it is sufficient to add `ole32` and `user32` to the system libraries?

traversaro commented:
Furthermore, the linux build is now failing due to: 
~~~
/mnt/vcpkg-ci/buildtrees/ignition-common3/src/mon3_3.5.0-8876f59078/include/ignition/common/Uuid.hh:31:12: fatal error: uuid/uuid.h: No such file or directory
   #include <uuid/uuid.h>
~~~
so probably the `uuid.pc` pkg-config file is malformed.

Neumann-A commented:
>do you think it is sufficient to add ole32 and user32 to the system libraries?

you can probably add all windows standard libraries as system libraries. 
From CMakes `Platform/Windows-MSVC.cmake`
```
elseif(WINDOWS_PHONE OR WINDOWS_STORE)
  set(_PLATFORM_DEFINES "/DWIN32")
  set(_FLAGS_C " /DUNICODE /D_UNICODE")
  set(_FLAGS_CXX " /DUNICODE /D_UNICODE /GR /EHsc")
  if(WINDOWS_STORE AND MSVC_VERSION GREATER 1899)
    set(CMAKE_C_STANDARD_LIBRARIES_INIT "WindowsApp.lib")
  elseif(WINDOWS_PHONE)
    set(CMAKE_C_STANDARD_LIBRARIES_INIT "WindowsPhoneCore.lib RuntimeObject.lib PhoneAppModelHost.lib")
  elseif(_MSVC_C_ARCHITECTURE_FAMILY STREQUAL "ARM" OR _MSVC_CXX_ARCHITECTURE_FAMILY STREQUAL "ARM" OR _MSVC_C_ARCHITECTURE_FAMILY STREQUAL "ARM64" OR _MSVC_CXX_ARCHITECTURE_FAMILY STREQUAL "ARM64")
    set(CMAKE_C_STANDARD_LIBRARIES_INIT "kernel32.lib user32.lib")
  else()
    set(CMAKE_C_STANDARD_LIBRARIES_INIT "kernel32.lib user32.lib gdi32.lib winspool.lib comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib uuid.lib odbc32.lib odbccp32.lib")
  endif()
else()
```
```
  if(_MSVC_C_ARCHITECTURE_FAMILY STREQUAL "ARM" OR _MSVC_CXX_ARCHITECTURE_FAMILY STREQUAL "ARM")
    set(CMAKE_C_STANDARD_LIBRARIES_INIT "kernel32.lib user32.lib")
  elseif(MSVC_VERSION GREATER 1310)
    if(CMAKE_VS_PLATFORM_TOOLSET MATCHES "v[0-9]+_clang_.*")
      # Clang/C2 in MSVC14 Update 1 seems to not support -fsantinize (yet?)
      # set(_RTC1 "-fsantinize=memory,safe-stack")
      set(_FLAGS_CXX " -frtti -fexceptions")
    else()
      set(_RTC1 "/RTC1")
      set(_FLAGS_CXX " /GR /EHsc")
    endif()
    set(CMAKE_C_STANDARD_LIBRARIES_INIT "kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib")
  else()
    set(_RTC1 "/GZ")
    set(_FLAGS_CXX " /GR /GX")
    set(CMAKE_C_STANDARD_LIBRARIES_INIT "kernel32.lib user32.lib gdi32.lib winspool.lib comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib uuid.lib odbc32.lib odbccp32.lib")
  endif()
```

traversaro commented:
> > do you think it is sufficient to add ole32 and user32 to the system libraries?
> 
> you can probably add all windows standard libraries as system libraries.

Thanks @Neumann-A ! The strange thing is that apparently the library are not listed as `-lole32`, but just as `ole32.lib`, see: 
~~~
Name: libavdevice
Description: FFmpeg device handling library
Version: 58.8.100
Requires: 
Requires.private: libavfilter >= 7.57.100, libswscale >= 5.5.100, libavformat >= 58.29.100, libavcodec >= 58.54.100, libswresample >= 3.5.100, libavresample >= 4.0.0, libavutil >= 56.31.100
Conflicts:
Libs: -L${libdir}  -lavdevice 
Libs.private: psapi.lib ole32.lib strmiids.lib uuid.lib oleaut32.lib shlwapi.lib gdi32.lib vfw32.lib
Cflags: -I${includedir}
~~~
do you think that is a case that is supported in pkg-config, or it is just that `ffmpeg` produces malformed `.pc` files?

traversaro commented:
I opened a separate issue for the ffmpeg/.pc files problem as it is not trivial: https://github.com/microsoft/vcpkg/issues/12376 .

traversaro commented:
> 
> 
> I opened a separate issue for the ffmpeg/.pc files problem as it is not trivial: #12376 .

According to @Neumann-A , that issue is blocked by the merging of https://github.com/microsoft/vcpkg/pull/11550 .

Furthermore, also on linux I  fixed some more issues (https://github.com/ignitionrobotics/ign-cmake/pull/104 and https://github.com/ignitionrobotics/ign-cmake/pull/105) and the build now fails with: 
~~~
CMake Error in graphics/src/CMakeLists.txt:
  Imported target "GTS::GTS" includes non-existent path

    "/home/straversaro/vcpkg/installed/x64-linux/debug/lib/pkgconfig/../../../include/glib-2.0"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.
~~~
that again seems something that will be solved by https://github.com/microsoft/vcpkg/pull/12326, that itself is  blocked by https://github.com/microsoft/vcpkg/pull/11550 and https://github.com/microsoft/vcpkg/pull/11550 .

traversaro commented:
> that again seems something that will be solved by #12326, that itself is blocked by #11550 and #11776 .

Status update: 
* https://github.com/microsoft/vcpkg/pull/11550 has been merged
* https://github.com/microsoft/vcpkg/pull/11776 still needs to be merged, blocked by https://github.com/microsoft/vcpkg/pull/11836 and https://github.com/microsoft/vcpkg/pull/12405
* https://github.com/microsoft/vcpkg/pull/12326 still needs to be merged, blocked by https://github.com/microsoft/vcpkg/pull/11776

traversaro commented:
> 
> 
> > that again seems something that will be solved by #12326, that itself is blocked by #11550 and #11776 .
> 
> Status update:
> 
>     * #11550 has been merged
> 
>     * #11776 still needs to be merged, blocked by #11836 and #12405
> 
>     * #12326 still needs to be merged, blocked by #11776

Status update:
* #11776 still needs to be merged, now blocked by https://github.com/microsoft/vcpkg/pull/12936 
* #12326 still needs to be merged, still blocked by #11776

traversaro commented:
> 
> 
> I opened a separate issue for the ffmpeg/.pc files problem as it is not trivial: #12376 .

This will be fixed (hopefully) by https://github.com/microsoft/vcpkg/pull/13919 .

traversaro commented:
Status update:
* #11776 still needs to be merged, still blocked by https://github.com/microsoft/vcpkg/pull/12936 
* #12326 still needs to be merged, still blocked by #11776 
* https://github.com/microsoft/vcpkg/pull/13919 still needs to be merged, possibly blocked by https://github.com/microsoft/vcpkg/pull/13126 and https://github.com/microsoft/vcpkg/pull/9966 depending on the outcome of the discussion in the PR

traversaro commented:
> 
> 
> Status update:
> 
>     * #11776 still needs to be merged, still blocked by #12936
> 
>     * #12326 still needs to be merged, still blocked by #11776
> 
>     * #13919 still needs to be merged, possibly blocked by #13126 and #9966 depending on the outcome of the discussion in the PR

New status update: 
* https://github.com/microsoft/vcpkg/pull/11776 merged! 
* https://github.com/microsoft/vcpkg/pull/12326 still needs to be merged, it is not blocked by any other PR
* https://github.com/microsoft/vcpkg/pull/13919 still needs to be merged, possibly blocked by https://github.com/microsoft/vcpkg/pull/13126 and https://github.com/microsoft/vcpkg/pull/9966 depending on the outcome of the discussion in the PR

traversaro commented:
Status updated:
* Cleanup the PR as several changes are not necessary anymore due to updated to ignition ports 
* Bumped used version to 3.9.0

> #13919 still needs to be merged, possibly blocked by #13126 and #9966 depending on the outcome of the discussion in the PR

This was superseded by https://github.com/microsoft/vcpkg/pull/15127 .

The Linux CI is now blocked by https://github.com/microsoft/vcpkg/issues/15326, that I guessed could be solved by https://github.com/microsoft/vcpkg/pull/12326 but probably can be solved also by other glib-related PRs.

traversaro commented:
Thanks to https://github.com/microsoft/vcpkg/pull/15360, the PR is now passing all the CI tests and is now ready for review. 

fyi @seanyen @ahoarau

PhoebeHui commented on ports/ignition-common3/portfile.cmake:
```diff
@@ -0,0 +1,12 @@
+                         SHA512 8d052850cbb125e334494c9ad9b234c371fe310327dba997515651f29479d747dffa55b0aa822f2a78e6317a4df2d41389c7a07165cdc08894fdfb116e4d9756)
+
+# Remove non-relocatable helper scripts (see https://github.com/ignitionrobotics/ign-common/issues/82)
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/bin/ign_remotery_vis" "${CURRENT_PACKAGES_DIR}/debug/bin'/ign_remotery_vis")
```
```suggestion
file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/bin/ign_remotery_vis" "${CURRENT_PACKAGES_DIR}/debug/bin/ign_remotery_vis")
```

traversaro replied:
Done!

traversaro committed:
Update ports/ignition-common3/portfile.cmake

Co-authored-by: Phoebe <20694052+PhoebeHui@users.noreply.github.com>

PhoebeHui reviewed (approved):
Thanks for your updates!

vicroms commented:
Thanks for the PR!
